core/src/main/scala/org/typelevel/paiges/Doc.scala

```diff
@@ -914,4 +954,14 @@ object Doc {
       def keyToDoc(s: String): Doc = Doc.text(s) + Doc.char(fill).repeat(fills - s.length) + rightD
       intercalate(line, rows.map { case (k, v) => keyToDoc(k) + v.aligned })
     }
+
+  private sealed trait FlattenResult extends Product with Serializable {
```

Comment:
That's fine with me, with one caveat.  

Overall, I'm torn about this change.  To me, copying Haskell's `Fail` constructor makes all of this more elegant.  On the other hand, it makes a number of functions like `renderWideStream` and `maxWidth` significantly more complex, requiring lookahead.  `fill` is also difficult, since it seems you made performance a priority there, and I don't have a good intuition of how the `Fail` version will perform, e.g. using the definition from the paper.    As such, I thought this was the minimal change that preserved the existing characteristics.  I'm not sure if it's the right decision or not.  

Making this public will mean that if we ever decide to use the `Fail` implementation, it will require a major version change.